README
:: Ranking ::

Assigning rank to each value.

Docs: https://ranking.readthedocs.io/

by Heungsub Lee


CHANGES
Version 0.3.2
-------------

Released on Oct 9th 2020.

- Stopped supporting Python 2.5, 2.6, 3.2, and 3.3.
- Supported Python 3.4, 3.5, 3.6, 3.7, and 3.8.

Thanks to `Nolan Gilley`_.

.. _Nolan Gilley: https://github.com/nkgilley

Version 0.3.1
-------------

Released on Mar 4th 2013.

Fixes an error of :func:`FRACTIONAL` strategy with multiple tie ranks. Thanks
to `Yunwon Jeong`_.

.. _Yunwon Jeong: http://youknowone.org/

Version 0.3
-----------

Released on Jan 30th 2013.

Doesn't support old ``cmp`` style just like Python 3.

- Removes``cmp`` parameter from :class:`Ranking`.
- Removes :func:`score_comparer`.
- ``key``, ``reverse``, ``no_score`` parameter of :class:`Ranking` is now
  keyword-only.

Version 0.2.4
-------------

Released on Jan 9th 2013.

Supports Python 3 and Jython.

Version 0.2.3
-------------

Released on Oct 26th 2012.

- Adds ``no_score`` parameter to use other value for "no score" instead of
  ``None``.
- Implements :func:`score_comparer`.

Version 0.2.2
-------------

Released on Oct 26th 2012.

- Adds ``reverse`` parameter for :class:`Ranking`.
- Adds :meth:`Ranking.rank` to find the rank of the value.

Version 0.2.2
-------------

Released on Oct 26th 2012.

- Adds ``reverse`` parameter for :class:`Ranking`.
- Adds :meth:`Ranking.rank` to find the rank of the value.

Version 0.1.2
-------------

Released on Oct 9th 2012.

- Uses ``None`` for "no score".
- Works with ``sequence`` as an iterator instead of only a list or tuple
  object.
- :class:`Ranking` yields values instead of found scores.

Version 0.1.1
-------------

Released on Oct 7th 2012.

- Adds ``start`` parameter for :class:`Ranking`.
- Renames ``score`` parameter for :class:`Ranking` to ``key``.

Version 0.1
-----------

First public preview release.


rankingtests.py
# -*- coding: utf-8 -*-
from __future__ import with_statement
from collections import defaultdict

from pytest import raises

from ranking import (Ranking, COMPETITION, MODIFIED_COMPETITION, DENSE,
                     ORDINAL, FRACTIONAL)


try:
    next
except NameError:
    # for Python 2.5
    next = lambda it: it.next()


def ranks(ranking):
    return list(ranking.ranks())


def test_competition():
    assert ranks(Ranking([5, 4, 4, 3], COMPETITION)) == [0, 1, 1, 3]


def test_modified_competition():
    assert ranks(Ranking([5, 4, 4, 3], MODIFIED_COMPETITION)) == [0, 2, 2, 3]


def test_dense():
    assert ranks(Ranking([5, 4, 4, 3], DENSE)) == [0, 1, 1, 2]


def test_ordinal():
    assert ranks(Ranking([5, 4, 4, 3], ORDINAL)) == [0, 1, 2, 3]


def test_fractional():
    assert ranks(Ranking([5, 4, 4, 3], FRACTIONAL)) == [0, 1.5, 1.5, 3]


def test_fractional_with_multiple_tie_ranks():
    assert ranks(Ranking([5, 4, 4, 4, 4, 3], FRACTIONAL)) == \
           [0, 2.5, 2.5, 2.5, 2.5, 5]


def test_unsorted():
    with raises(ValueError):
        list(Ranking([5, 4, 4, 5]))


def test_strategies():
    strategies = [COMPETITION, MODIFIED_COMPETITION, DENSE, ORDINAL,
                  FRACTIONAL]
    for strategy in strategies:
        assert len(list(strategy(0, 2))) == 3


def test_capsuled_scores():
    class User(object):
        def __init__(self, score):
            self.score = score
        def __lt__(self, other):
            raise NotImplemented
        def __gt__(self, other):
            raise NotImplemented
    users = [User(100), User(80), User(80), User(79)]
    with raises(TypeError):
        list(Ranking(users))
    key = lambda user: user.score
    ranking = Ranking(users, key=key)
    assert ranks(ranking) == [0, 1, 1, 3]
    assert isinstance(next(iter(ranking))[1], User)


def test_less_is_more():
    records = [1, 121, 121, 432, None, None]
    with raises(ValueError):
        list(Ranking(records))
    assert ranks(Ranking(records, reverse=True)) == [0, 1, 1, 3, None, None]


def test_empty():
    assert list(Ranking([])) == []
    with raises(TypeError):
        Ranking()


def test_start_from_not_zero():
    assert ranks(Ranking([5, 4, 4, 3], start=10)) == [10, 11, 11, 13]


def test_iterator_aware():
    scores = (x for x in range(100, 50, -10))
    assert ranks(Ranking(scores)) == [0, 1, 2, 3, 4]


def test_no_score_no_rank():
    assert ranks(Ranking([100, 50, 50, None, None])) == [0, 1, 1, None, None]
    assert ranks(Ranking([None])) == [None]
    assert ranks(Ranking([None, None])) == [None, None]
    assert ranks(Ranking([3, 1, 1, None])) == [0, 1, 1, None]


def test_custom_no_score():
    assert ranks(Ranking([100, 50, 50, -1, -1], no_score=-1)) == \
           [0, 1, 1, None, None]
    assert ranks(Ranking([-1], no_score=-1)) == [None]
    assert ranks(Ranking([-1, -1], no_score=-1)) == [None, None]
    assert ranks(Ranking([3, 1, 1, -1], no_score=-1)) == [0, 1, 1, None]
    assert ranks(Ranking([1, 1, 3, -1], reverse=True, no_score=-1)) == \
           [0, 0, 2, None]


def test_multiple_ties():
    assert ranks(Ranking([5, 5, 5, 3, 3, 3, 2, 2, 1, 1, 1, 1])) == \
           [0, 0, 0, 3, 3, 3, 6, 6, 8, 8, 8, 8]


def test_custom_strategy():
    def exclusive(start, length):
        return [None] * length + [start]
    assert ranks(Ranking([100, 80, 80, 70], exclusive)) == [0, None, None, 1]


def test_keyword_only():
    with raises(TypeError):
        Ranking([], COMPETITION, 0, None)


def test_no_cmp():
    with raises(TypeError):
        Ranking([], cmp=lambda a, b: -1)


ranking.py
# -*- coding: utf-8 -*-
"""
    ranking
    ~~~~~~~

    :class:`Ranking` and various strategies for assigning rankings.

    :copyright: (c) 2012-2020 by Heungsub Lee
    :license: BSD, see LICENSE for more details.
"""
import itertools


__version__ = '0.3.2'
__all__ = ['Ranking', 'COMPETITION', 'MODIFIED_COMPETITION', 'DENSE',
           'ORDINAL', 'FRACTIONAL']


try:
    next
except NameError:
    # for Python 2.5
    next = lambda it: it.next()


def strategy(f):
    class Strategy(object):
        def __call__(self, start, length):
            return f(start, length)
        def __repr__(self):
            return f.__name__
    Strategy.__name__ = f.__name__
    return Strategy()


@strategy
def COMPETITION(start, length):
    """Standard competition ranking ("1224" ranking)"""
    for x in xrange(length):
        yield start
    yield start + length


@strategy
def MODIFIED_COMPETITION(start, length):
    """Modified competition ranking ("1334" ranking)"""
    for x in xrange(length):
        yield start + length - 1
    yield start + length


@strategy
def DENSE(start, length):
    """Dense ranking ("1223" ranking)"""
    for x in xrange(length):
        yield start
    yield start + 1


@strategy
def ORDINAL(start, length):
    """Ordinal ranking ("1234" ranking)"""
    return xrange(start, start + length + 1)


@strategy
def FRACTIONAL(start, length):
    """Fractional ranking ("1 2.5 2.5 4" ranking)"""
    avg = (2 * start + length - 1) / 2.
    for x in xrange(length):
        yield avg
    yield start + length


class Ranking(object):
    """This class looks like `enumerate` but generates a `tuple` containing a
    rank and value instead.

    >>> scores = [100, 80, 80, 70, None]
    >>> list(Ranking(scores))
    [(0, 100), (1, 80), (1, 80), (3, 70), (None, None)]

    :param sequence: sorted score sequence. `None` in the sequence means that
                     no score.
    :param strategy: a strategy for assigning rankings. Defaults to
                     :func:`COMPETITION`.
    :param start: a first rank. Defaults to 0.
    :param key: (keyword-only) a function to get score from a value
    :param reverse: (keyword-only) `sequence` is in ascending order if `True`,
                    descending otherwise. Defaults to `False`.
    :param no_score: (keyword-only) a value for representing "no score".
                     Defaults to `None`.
    """

    def __init__(self, sequence, strategy=COMPETITION, start=0, **kwargs):
        self.sequence = sequence
        self.strategy = strategy
        self.start = start
        self.key = kwargs.pop('key', None)
        self.reverse = kwargs.pop('reverse', False)
        self.no_score = kwargs.pop('no_score', None)
        if kwargs:
            s = 's' if len(kwargs) == 2 else ''
            params = ', '.join(repr(key) for key in kwargs.iterkeys())
            raise TypeError('Unexpected keyword argument%s: %s' % (s, params))

    def __iter__(self):
        rank, drawn, tie_started, final = self.start, [], None, object()
        iterator = iter(self.sequence)
        try:
            right = next(iterator)
        except StopIteration:
            return
        right_score = right if self.key is None else self.key(right)
        for value in itertools.chain(iterator, [final]):
            left, right = right, value
            left_score = right_score
            if value is not final:
                right_score = right if self.key is None else self.key(right)
            if left_score == self.no_score:
                yield None, left
                continue
            elif right_score == self.no_score or value is final:
                compared = 1
            else:
                compared = (0 if left_score == right_score else
                            -1 if left_score < right_score else 1)
                if self.reverse:
                    compared = -compared
            if compared < 0: # left is less than right
                raise ValueError('%r should be behind of %r' % \
                                 (left_score, right_score))
            elif compared == 0: # same scores
                if tie_started is None:
                    tie_started = rank
                drawn.append(left)
                continue
            elif drawn:
                drawn.append(left)
                for rank in self.strategy(tie_started, len(drawn)):
                    try:
                        yield rank, drawn.pop(0)
                    except IndexError:
                        pass
                tie_started = None
                continue
            yield rank, left
            rank += 1

    def ranks(self):
        """Generates only ranks."""
        for rank, value in self:
            yield rank

    def rank(self, value):
        """Finds the rank of the value.

        :raises ValueError: the value isn't ranked in the ranking
        """
        for rank, other in self:
            if value == other:
                return rank
        raise ValueError('%r is not ranked' % value)
